### PR TITLE
feat(native-filters): Support default to first value in select filter

### PR DESCRIPTION
**Pull request by simcha90**
### SUMMARY

This PR fixes support for default first value in Select native filter. It includes next logic:
- if one of control panel items contains `requiredFirst` flag we cancel delay of native filters loading, delay charts loading until this filter will be initialized - we do assumption if `filterState.value === undefined`, it means that filter not initialized
- if it has `defaultToFirstValue` option enabled - we save on `filterState.value` undefined value and on filter loading we on the fly put first value

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

https://user-images.githubusercontent.com/56388545/119837538-5b2aa900-bf0b-11eb-9eee-83c90402477d.mov

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by villebro** [comment]
@simcha90 this needs a rebase. Let's also break out the Select filter tests into a separate PR to make review easier.

File: superset-frontend/src/dashboard/components/DashboardBuilder/DashboardBuilder.tsx [hunk 1]
```diff
@@ -49,9 +48,9 @@ import {
 import FilterBar from 'src/dashboard/components/nativeFilters/FilterBar';
 import { StickyVerticalBar } from '../StickyVerticalBar';
 import { shouldFocusTabs, getRootLevelTabsComponent } from './utils';
-import { useFilters } from '../nativeFilters/FilterBar/state';
-import { Filter } from '../nativeFilters/types';
 import DashboardContainer from './DashboardContainer';
+import Loading from '../../../components/Loading';
```
Comment (villebro) [inline superset-frontend/src/dashboard/components/DashboardBuilder/DashboardBuilder.tsx:52]: nit
```
import Loading from 's/components/Loading';
```

File: superset-frontend/src/dashboard/components/DashboardBuilder/DashboardBuilder.tsx [hunk 2]
```diff
@@ -159,15 +138,12 @@ const DashboardBuilder: FC<DashboardBuilderProps> = () => {
...
-    ) {
-      toggleDashboardFiltersOpen(false);
-    }
-  }, [filterValues.length]);
+  const {
+    showDashboard,
+    dashboardFiltersOpen,
+    toggleDashboardFiltersOpen,
+    nativeFiltersEnabled,
+  } = useNativeFilters();
```
Comment (villebro) [inline superset-frontend/src/dashboard/components/DashboardBuilder/DashboardBuilder.tsx:146]: good idea to split out into own hook 👍

File: superset-frontend/src/dashboard/components/DashboardBuilder/state.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    ({ requiredFirst }) => requiredFirst,
+  );
+  const dataMask = useNativeFiltersDataMask();
+  const showDashboard = isInitialized
+    ? true
+    : !nativeFiltersEnabled ||
+      !(
+        nativeFiltersEnabled &&
+        requiredFirstFilter.length &&
+        requiredFirstFilter.find(
```
Comment (villebro) [inline superset-frontend/src/dashboard/components/DashboardBuilder/state.ts:58]: This seems simpler
```suggestion
  const showDashboard =
    isInitialized ||
    !nativeFiltersEnabled ||
    !(
      nativeFiltersEnabled &&
      requiredFirstFilter.length &&
      requiredFirstFilter.find(
        ({ id }) => dataMask[id]?.filterState?.value === undefined,
      )
    );
```

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/DefaultValue.tsx [hunk 4]
```diff
@@ -62,7 +62,7 @@ const DefaultValue: FC<DefaultValueProps> = ({
   ) : (
     <SuperChart
       height={25}
-      width={250}
+      width={180}
```
Comment (villebro) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/DefaultValue.tsx:65]: This feels slightly tight:
image
Maybe we should leave this change to a separate PR if it's needed so we can concentrate on the design aspects there and make this PR focus only on functional fixes related to the first value option?

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx [hunk 5]
```diff
@@ -599,9 +601,11 @@ const FiltersConfigForm = (
...
                         return Promise.resolve();
                       }
-                      return Promise.reject(
-                        new Error(t('Default value is required')),
-                      );
+                      return Promise.resolve();
+                      // TODO: Some values does default value empty and disabled so there is no value
+                      // return Promise.reject(
+                      //   new Error(t('Default value is required')),
+                      // );
```
Comment (villebro) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx:608]: I don't think we need to make this change - AFAIK it's enough to set the "default to first value" option and save, so no need to check the "default value" checkbox. I'd rather state in the description tooltip of the "default to first value" option that when using this option, setting a default value is not possible.

File: superset-frontend/src/filters/components/Select/SelectFilterPlugin.tsx [hunk 6]
```diff
@@ -16,6 +16,7 @@
  * specific language governing permissions and limitations
  * under the License.
  */
+/* eslint-disable no-param-reassign */
```
Comment (villebro) [inline superset-frontend/src/filters/components/Select/SelectFilterPlugin.tsx:19]: thought: if we want to encourage use of `useImmerReducer`, we should probably disable this linting rule.

**Review by villebro — COMMENTED** [review]
Some comments

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/DefaultValue.tsx (hunk 4 above)
Comment (simcha90) [reply superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/DefaultValue.tsx:65]: I did it because of in Native filters modal it's too large, but I can cancel it here

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx (hunk 5 above)
Comment (simcha90) [reply superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx:608]: current logic of defaultToFirstValue, requires to leave default value as `undefined` so it will fail on this validation, we can put some specific value here because it can different for different users

File: superset-frontend/src/filters/components/Select/SelectFilterPlugin.tsx (hunk 6 above)
Comment (simcha90) [reply superset-frontend/src/filters/components/Select/SelectFilterPlugin.tsx:19]: not sure, if we need it in all places disabled... only where we use immer, so may be later

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/index.tsx [hunk 7]
```diff
@@ -178,10 +174,21 @@ const FilterBar: React.FC<FiltersBarProps> = ({
 
   const handleFilterSelectionChange = (
     filter: Pick<Filter, 'id'> & Partial<Filter>,
-    dataMask: Partial<DataMaskState>,
+    dataMask: Partial<DataMask>,
   ) => {
     setIsFilterSetChanged(tab !== TabIds.AllFilters);
     setDataMaskSelected(draft => {
+      // force instant updating on initialization or for parent filters when dataMaskSelected has filter
```
Comment (villebro) [inline superset-frontend/src/dashboard/components/nativeFilters/FilterBar/index.tsx:181]: This comment is not fully accurate; can we update this to reflect the fact that we are updating on initialization if `filter.requiredFirst` is true?

File: superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx [hunk 8]
```diff
@@ -625,7 +627,11 @@ const FiltersConfigForm = (
                   {
                     validator: (rule, value) => {
                       const hasValue = !!value.filterState?.value;
-                      if (hasValue) {
+                      if (
+                        hasValue ||
+                        // TODO: do more generic
+                        formFilter.controlValues?.defaultToFirstItem
+                      ) {
```
Comment (villebro) [inline superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx:634]: I'd recommend leaving this unchanged and just state in the description of `defaultToFirstItem` that the default value will be populated automatically, thus default value can't be set when `defaultToFirstItem` is set to `true`.

**Review by villebro — COMMENTED** [review]
Two minor comments

File: superset-frontend/src/dashboard/components/nativeFilters/FilterBar/index.tsx (hunk 7 above)
Comment (simcha90) [reply superset-frontend/src/dashboard/components/nativeFilters/FilterBar/index.tsx:181]: it happens not only on initialisation - it also happens when we update filter modal, because first it set as undefined and then get correct first value (meantime it still do double loading but need fix later who will have time)

File: superset-frontend/src/filters/components/Select/controlPanel.ts [hunk 9]
```diff
@@ -78,7 +78,7 @@ const config: ControlPanelConfig = {
               renderTrigger: true,
               description: t(
                 'User must select a value for this filter when filter is in single select mode. ' +
-                  'If selection is empty, an always false filter is emitted.',
+                  'If selection is empty, an always false filter is emitted (when using this option, default value can’t be set)',
```
Comment (villebro) [inline superset-frontend/src/filters/components/Select/controlPanel.ts:81]: I don't think this option needs to be updated - did you mean to update `defaultToFirstItem` below?

File: superset-frontend/src/filters/components/Select/controlPanel.ts [hunk 10]
```diff
@@ -93,7 +93,10 @@ const config: ControlPanelConfig = {
               resetConfig: true,
               affectsDataMask: true,
               renderTrigger: true,
-              description: t('Select first item by default'),
+              requiredFirst: true,
+              description: t(
+                'Select first item by default (requires repopulate Default Value if column field changed)',
```
Comment (villebro) [inline superset-frontend/src/filters/components/Select/controlPanel.ts:98]: I think the text in parenthesis can be replaced with that above (we don't need to repopulate Default Value anymore).

**Review by villebro — APPROVED** [review]
LGTM
